dataset/src/main/scala/frameless/functions/NonAggregateFunctions.scala

```diff
@@ -176,107 +166,99 @@ trait NonAggregateFunctions {
     *
     * apache/spark
     */
-  def instr[T](column: TypedColumn[T, String], substring: String): TypedColumn[T, Int] = {
-    new TypedColumn[T, Int](untyped.instr(column.untyped, substring))
-  }
+  def instr[T](str: AbstractTypedColumn[T, String], substring: String): str.ThisType[T, Int] =
+    str.typed(untyped.instr(str.untyped, substring))
 
   /** Non-Aggregate function: Computes the length of a given string.
     *
     * apache/spark
     */
   //TODO: Also for binary
-  def length[T](column: TypedColumn[T, String]): TypedColumn[T, Int] = {
-    new TypedColumn[T, Int](untyped.length(column.untyped))
-  }
+  def length[T](str: AbstractTypedColumn[T, String]): str.ThisType[T, Int] =
+    str.typed(untyped.length(str.untyped))
 
   /** Non-Aggregate function: Computes the Levenshtein distance of the two given string columns.
     *
     * apache/spark
     */
-  def levenshtein[T](l: TypedColumn[T, String], r: TypedColumn[T, String]): TypedColumn[T, Int] = {
-    new TypedColumn[T, Int](untyped.levenshtein(l.untyped, r.untyped))
-  }
+  def levenshtein[T](l: AbstractTypedColumn[T, String], r: AbstractTypedColumn[T, String]): l.ThisType[T, Int] =
```

Comment:
Why not `r.ThisType`? Does this even make sense when `l.ThisType` != `r.ThisType`?

Reply:
Same question for `concat` and `concatWs`

Reply:
I thought found an easy solution ... was kind of surprised this doesn't work ...
```scala
def concatWs[T, G[_,_] <: AbstractTypedColumn[T, String]](sep: String,
                  c1: G[T, String],
                  rest: G[T, String]*): c1.ThisType[T, String] =
    c1.typed(untyped.concat_ws(sep, (c1 +: rest).map(_.untyped): _*))
```

Reply:
`G[_,_] <: AbstractTypedColumn[T, String]` doesn't mean what you think it means. Here you wrote, Some `G` that takes two (unnamed) arguments which is a subtype of `AbstractTypedColumn[T, String]` (no matter how you instantiate it's arguments.

So the following `G` would pass the typechecker:

```
class G[X, Y] extends AbstractTypedColumn[Int, String] // assuming T == Int, note that X and Y are *not* used anywhere, just declared.
```

What you want to express is that this method require some `G` taking two arguments `A` and `B`, such that `G` is a subtype of `AbstractTypedColumn` when applied to those *same two arguments*. The convention is to use lower case named for such locally instantiated type parameters:

```scala
def concatWs[T, G[a, b] <: AbstractTypedColumn[a, b]](...)
```

Reply:
Even with `-Ypartial-unification` scalac can't infer these types :(

Reply:
Never seen this done this way. Nice :).

Yeah, even doing this, it doesn't really force all Gs to be of the same type. It would still compile when you have one being aggr and one being projected..

Reply:
Well in theory it should do the right thing, but type inference can't handle the `[_, _]` here. This is the signature I have for levenshtein:

```scala
  def levenshtein[T, ColumnType[a, b] <: AbstractTypedColumn[a, b]](
    l: AbstractTypedColumn[T, String] { type ThisType[a, b] = ColumnType[a, b] },
    r: AbstractTypedColumn[T, String] { type ThisType[a, b] = ColumnType[a, b] }
  ): ColumnType[T, Int] =
    l.typed(untyped.levenshtein(l.untyped, r.untyped))
```

I think we can just go for the dirty solution: duplicate all the methods involving two `AbstractTypedColumn`.

Reply:
OhhI think I got it(?)

Usage:
```scala
scala> val t1 = TypedDataset.create((1,"b")::(2,"c")::Nil)
t1: frameless.TypedDataset[(Int, String)] = [_1: int, _2: string]

scala> import frameless.functions.nonAggregate._
import frameless.functions.nonAggregate._

scala> t1.agg(concatWs(",", sum(t1('_1)).cast[String], t1('_2))).show().run
<console>:29: error: Cannot prove that frameless.TypedAggregate[_, _] =:= frameless.TypedColumn[_, _].
       t1.agg(concatWs(",", sum(t1('_1)).cast[String], t1('_2))).show().run
                      ^

scala> t1.agg(concatWs(",", sum(t1('_1)).cast[String], sum(t1('_1)).cast[String])).show().run
+---+
| _1|
+---+
|3,3|
+---+
```

Here is what seemed to work:

```scala
 def concatWs[T, G1[a,b] <: AbstractTypedColumn[a, b], G2[a,b] <: AbstractTypedColumn[a, b]](sep: String,
                  c1: G1[T, String],
                  rest: G2[T, String]*)(implicit eq: G1[_,_] =:= G2[_,_]): c1.ThisType[T, String] =
    c1.typed(untyped.concat_ws(sep, (c1 +: rest).map(_.untyped): _*))
```

Reply:
Nice :smile:. But I'm not sure it's worth the generality, see my latest commit where I duplicate these methods...